### ISSUE TITLE
2.13 scala.util.Using: surprising behavior in for-comprehensions, `.map` method is non-compositional

### ISSUE THREAD

**Issue by kaishh** [scala/bug#11225]
Example:

```scala
import scala.util._

class Res() extends AutoCloseable {
  var open = true

  override def close() = open = false

  def use() = {
    if (!open) throw new Exception("Already closed!")
  }
}

object App extends App {
  val startTime = System.nanoTime
  (for {
    resource <- Using(new Res())
    endTime = System.nanoTime
  } yield {
    println(s"initialization took: ${endTime-startTime} ns")
    resource.use()
  }).get
}
```

Expected `resource.use()` to succeed

Actual: Got exception `Already closed!`

Cause: `resource` was closed right at `endTime = System.nanoTime` instead of extending until after `yield`. If `endTime` is written as `endTime <- Try(System.nanoTime)` the behavior is as expected.

**Comment by NthPortal** [comment]
The problem probably also happens if you use a filter

**Comment by SethTisue** [comment]
is it a regression?

**Comment by kaishh** [comment]
@SethTisue No, Using is a new feature in 2.13 standard library https://github.com/scala/scala/commit/97f2a857f4af03d20d0aa8649f9ff0d0e84bc0e2

**Comment by NthPortal** [comment]
@SethTisue I think it's going to require a non-trivial redesign of the "safe" version of `Using`. I'm currently looking into what better-files does, to see if it's useful for this

**Comment by NthPortal** [comment]
@Ichoran do you have any thoughts on how to solve this problem? Since a lot of the design was yours

**Comment by NthPortal** [comment]
oops! misclick

(the number of times I have accidentally closed an issue because the button is the same colour and in the same location as the "Cancel" button for PR comments is really high)

**Comment by NthPortal** [comment]
I'm struggling to come up with a good solution to this. It seems like, due to the way `for` comprehensions are desugared, computation *cannot* take place during the `map` call. This yields the not terribly satisfying result of having `map` return a `Using`, and having a `use()` method on it to actually do the computation and get a result.

I'm not super happy with the usability, but I don't really have any other ideas

```scala
// old way
val res1 = for {
  reader <- Using(...)
} yield reader.readLines()

// new way

val res2 = (for {
  reader <- Using(...)
} yield reader.readLines()).use()
```

**Comment by NthPortal** [comment]
Possible new API/impl: https://gist.github.com/NthPortal/65ec5938fcb02c9c724d88b3b929e372

**Comment by kaishh** [comment]
Map desugaring is left-associative, so yeah, there's no way around other than suspending the computation. Alternatively, we could have only a .map method, without .withFilter, that returns e.g. a wrapper around Try *without* a .map method. That way a user wouldn't be able to chain .maps and the only .map use will be the final Yield.

**Comment by NthPortal** [comment]
My only concern is that if the result is a wrapped `Try` (let's call it `Used`, as a placeholder), how do you easily unwrap `Used`? If it's an implicit conversion, that could sneak into the middle of a `for` comprehension` and get you the wrong behaviour again.

**Comment by NthPortal** [comment]
On a more overarching note, I think that if we can't come up with a good solution in a reasonable amount of time before RC1, we should drop the `for` comprehension methods and just leave it with `apply` for 2.13.

**Comment by NthPortal** [comment]
One thing I thought of is that we can define `map` and `withFilter` on `Used`, but with an impossible implicit parameter (`Nothing`?). Then, we can use `@implicitNotFound` to tell users what to do instead. Additionally, we can then use an implicit conversion to `Try`, since that won't be used by the compiler in a `for` comprehension.

Thoughts?

**Comment by kaishh** [comment]
Even with a bad implicit it will still be callable with .asInstanceOf or `implicit def a[A]: A` IMHO it's better to just not have the offending methods

**Comment by NthPortal** [comment]
@kaishh true, but then they're very deliberately shooting themselves in the foot. There's no real way to prevent that.

**Comment by kaishh** [comment]
I'm just saying that there would be more invalid states, like, you'd also have to think about a runtime error message, whereas 'no method' means just one thing - 'no method'

**Comment by Jasper-M** [comment]
I don't think there's a clean way out of explicitly calling `use()`. Everything else seems hacky and confusing...

**Comment by NthPortal** [comment]
I think for now it would be best to get rid of the problematic methods. If anyone comes up with a good way to have them work, that would be great and we can add them back (before release).
